Documentation of plugin development and usage (#9)
STORM-IRIT/Radium-PluginExample: README.md

cmake --build .
```

When configuring the examples, the cmake command takes the following options :

-   `Radium_DIR` must be set to `pathToRadiumInstallation/lib/cmake/Radium`
-   `INSTALL_IN_RADIUM_BUNDLE`, Off by default.

By default the install target installs the plugin into the directory `installed-<Compiler_id>` in the plugin buildtree. Optional cmake options `-DCMAKE_INSTALL_PREFIX=pathToInstallPlugins`  or `-DINSTALL_IN_RADIUM_BUNDLE=ON` can be provided to install the plugins and resources in specified directory or in the Radium-Engine installation directory. Plugins installed into the Radium-Engine installation directory are automatically loaded by compatible applications while other installation directory must explicitly be added to the plugin directories of the applications.

## Description of the example plugins

For the prodvided plugins, the cmake option `INSTALL_IN_RADIUM` allows to add the parameter `INSTALL_IN_RADIUM_BUNDLE`
to each plugin CMakeLists.txt so that the Example plugins will be isntalled into the Radium Bundle
(i.e. use `cmake -DINSTALL_IN_RADIUM=On`). Note that this option requires to have write access to the Radium Bundle 

> installation directory.

Reviewer: This should be merged into the above section

> installation directory.

Reviewer: This should be merged into the above section

It is developed and maintained by the [STORM research group](https://www.irit.fr/STORM/site/).

This repository holds examples of Plugins based on Radium Libraries.

> To compile theses examples, you first need to compile the [Radium libraries](https://github.com/STORM-IRIT/Radium-Engine) and an application based Radium, for instance [Main Application](https://github.com/STORM-IRIT/Radium-Apps).

Author: ```suggestion
To compile theses examples, you first need to compile the [Radium libraries](https://github.com/STORM-IRIT/Radium-Engine) and a plugin compatible application based on Radium, for instance [Main Application](https://github.com/STORM-IRIT/Radium-Apps).
```

When configuring the examples, the cmake command takes the following options :
 - `Radium_DIR` must be set to `pathToRadiumInstallation/lib/cmake/Radium`
 - `INSTALL_IN_RADIUM_BUNDLE`, Off by default.

These options are described [below](#configure-the-plugin)

> **TD;DR;**

Author: ```suggestion
```

```Shell
mkdir build-release && cd build-release

> cmake ../ -DRadium_DIR=pathToRadiumInstallation/lib/cmake/Radium -DINSTALL_IN_RADIUM_BUNDLE=ON -DCMAKE_BUILD_TYPE=Release

Author: I'm not sure that the option `-DINSTALL_IN_RADIUM_BUNDLE=ON` might be written as this in the exemple. Installing directly Plugins into the bundle is not a good idea as it will complicate the bundle maintenance and many peaople will just copy-paste the given exemple
```suggestion
cmake ../ -DRadium_DIR=pathToRadiumInstallation/lib/cmake/Radium  -DCMAKE_BUILD_TYPE=Release [-DCMAKE_INSTALL_PREFIX=pathToInstallPlugins | -DINSTALL_IN_RADIUM_BUNDLE=ON]
```

Reviewer 2: Ok, so I still don't get the fast/correct way of making the main-app working with plugins.
I compile and install radium, compile main-app with -DRadium_DIR
compile plugins (and install them in their build tree ?) 
then I run the main-app and search for every plugin install dir ? by hand ?

Author: There is three use cases :
1. You are developing/testing your plugin.
  DO NOT INSTALL the plugin, just build it, and tell main-app where to find it in its build tree (e.g `build-release/Plugins`). You have to do this ONCE for each plugin. Each time you will launch main-app, it will search in ALL its registered location.
  The advantage is a fast startup of the application (as you do not install plugins) and efficient update of the Plugin resources as Resources directories are linked to the build tree.
2. You have developed several plugins for your use of the main-app and you will only occasionally recompile your plugins.
   in this case, install your plugin in ONE UNIQUE LOCATION (where you want in you system) and tell the application to load plugins here (e.g. make install...). To do that, use the option `-DCMAKE_INSTALL_PREFIX=MY_UNIQUE_LOCATION` (with hopefully the same `MY_UNIQUE_LOCATION` for each plugin). Each time the application will start, ALL the plugins from this `MY_UNIQUE_LOCATION` will be loaded, even those installed here after the configuration of the plugin location in the app.
3. You want to distribute plugins BINARIES with Radium BINARIES
  in this case, use the `INSTALL_IN_RADIUM_BUNDLE` option.

The use case 1 is efficient when you develop/test plugins with resources, the use case 2 is efficient when you have mature plugins you want to stack together, the 3 is fine when you want to distribute radium as binaries.
But, if you want to do things another-way, you can. The documentation has just to explain the main usage of plugins, mainly for development as it is a developper documentation. I'll put the description of the 3 above use cases in the documentation to be more clear.

-   FakeToonMaterial: illustrates how to add a new material in Radium. All the models loaded in MainApplication when this plugin is enabled are assigned with a toon shader.
-   CameraManip: illustrates how to create cameras in Radium and switch from one to another.
-   Dummy: illustrates how to create a new system to add components to entities at model load, and tasks to the engine at each frame.
-   MeshFeatureTracking: illustrates how to pick mesh features (vertex, edge, triangle) and display data about these, updated at each frame.
-   MeshPaint: illustrates how to paint on a mesh, here painting the per-vertex color.

## Quick instruction to build and install the examples

```Shell
mkdir build-release && cd build-release
cmake ../ -DRadium_DIR=pathToRadiumInstallation/lib/cmake/Radium -DCMAKE_BUILD_TYPE=Release
cmake --build .
```

When configuring the examples, the cmake command takes the following options :

-   `Radium_DIR` must be set to `pathToRadiumInstallation/lib/cmake/Radium`
-   `INSTALL_IN_RADIUM_BUNDLE`, Off by default.

> By default the install target installs the plugin in `installed-<Compiler_id>` in the plugin buildtree. Optional cmake options `-DCMAKE_INSTALL_PREFIX=pathToInstallPlugins`  or `-DINSTALL_IN_RADIUM_BUNDLE=ON` can be provided to install the plugins and resources in  specified diretory or in the Radium-Engine installation directory. Plugins installed Radium-Engine installation directory are autmatically loaded by compatible applicaitions while other installation directory must explicitly be added to the plugin directories of the applications.

Author: ```suggestion
By default the install target installs the plugin into the directory `installed-<Compiler_id>` in the plugin buildtree. Optional cmake options `-DCMAKE_INSTALL_PREFIX=pathToInstallPlugins`  or `-DINSTALL_IN_RADIUM_BUNDLE=ON` can be provided to install the plugins and resources in specified directory or in the Radium-Engine installation directory. Plugins installed into the Radium-Engine installation directory are automatically loaded by compatible applications while other installation directory must explicitly be added to the plugin directories of the applications.
```

This installed plugin is then relocatable only on the system where the plugin was built. 
You can change this default behavior by setting `CMAKE_INSTALL_PREFIX`. To install
the plugin in `pathToRadiumInstallation/Plugins/lib/`, which is relocatable alongside the Radium bundle and 
automatically loaded by any Plugin-aware applications (such as `main-app`), pass the parameter `INSTALL_IN_RADIUM_BUNDLE`
to the `configure_radium_plugin_install` of each plugin's `CMakeLists.txt`. 

> For the prodvided plugins, the option `INSTALL_IN_RADIUM` allows to add the parameter for each plugins (i.e. use `cmake -DINSTALL_IN_RADIUM=On`).

Reviewer 3: I don't understand this sentence

Plugins could be loaded and used from their build-trees.

### Configure the plugin

> A general advice is to follow out of source build. For instance:

Author: the configuration of the cmake prevent to build in source. So, out of source building is mandatory.
see cmake commands 
```
set(CMAKE_DISABLE_SOURCE_CHANGES ON)
set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)
```
located at the beginning of the cmake exemple below in the doc (and in all the CMakeLists.txt in the plugin example source tree).
```suggestion
As in source build is not recommended (and even forbidden by the CMakeLists.txt example configuration) a general advice is to follow out of source build. For instance:
```

    and efficient update of the Plugin resources as Resources directories are linked (symbolic links on systems that 
    support them) into the build tree.

2.  You have developed several plugins, and will only occasionally recompile them, for your own usage and you want your 
    plugin compatible applications to find and load them quickly.
    In this case, you might install your plugin in one unique location and tell the applications to load plugins from 
    there. To do so, use the option `-DCMAKE_INSTALL_PREFIX=thePathToTheInstalledPlugins` and do a `make install`
    You can then configure your applications so that they will search plugins from the `thePathToTheInstalledPlugins` directory.
    Each time the application will start, all the plugins from this `thePathToTheInstalledPlugins` will be 
    loaded, even those installed here after the configuration of the plugin location in the application.

3.  You want to distribute plugins binaries alongside the Radium bundle.
    In this case, you have to configure the plugin with the`INSTALL_IN_RADIUM_BUNDLE`option and do a `make install` to

>     install the plgins in the Radium Bundle. Note that this requires write access to the Radium Bundle.

Author: ```suggestion
    install the plugins in the Radium Bundle. Note that this requires write access to the Radium Bundle.
```

cmake --build .
```

If you just want to use the plugin directly from its build-tree, the resulting dynamic library will be found in the 
`build-release/Plugins/lib` directory and the associated resources will be linked (or copied on systems that do not 
support symbolic links) in the directory `build-release/Plugins/Resources`. In this case, note that the plugin from the 
buildtree could generally not be moved as it is linked to used libraries using relative paths.

If you install the plugin, by default, it will be installed in a directory
called `installed-<Compiler_id>` in the plugin buildtree. 
This installed plugin is then relocatable only on the system where the plugin was built.  
You can change this default behavior by setting `CMAKE_INSTALL_PREFIX` to any directory, e.g `pathToInstallPlugins`.

> The dyhnamic library will then be copied into  `pathToInstallPlugins/lib` and the Resources will be copied into 

Author: ```suggestion
The dynamic library will then be copied into  `pathToInstallPlugins/lib` and the Resources will be copied into 
```
